runtime/src/main/java/org/corfudb/recovery/FastSmrMapsLoader.java

```diff
@@ -0,0 +1,328 @@
+     * @param end end address for the bulk read
+     * @return logData map ordered by addresses (increasing)
+     */
+    private Map<Long, ILogData> getLogData(long start, long end) {
+        return runtime.getAddressSpaceView().
+                read(ContiguousSet.create(Range.closedOpen(start, end), DiscreteDomain.longs()));
+    }
+
+
+    /** Create a new object SMRMap as recipient of SMRUpdates (if doesn't exist yet)
+     *
+     * @param streamId
+     */
+    private void createSmrMapIfNotExist(UUID streamId) {
+        if (!runtime.getObjectsView().getObjectCache().containsKey(getObjectIdFromStreamID(streamId))) {
+            runtime.getObjectsView().build()
+                    .setStreamID(streamId)
+                    .setType(SMRMap.class)
+                    .open();
+        }
+    }
+
+    /** Deserialize a logData by getting the logEntry
+     *
+     * Getting the underlying logEntry should trigger deserialization only once.
+     * Next access should just returned the logEntry direclty.
+     *
+     * @param logData
+     * @return
+     * @throws Exception
+     */
+    public LogEntry deserializeLogData(ILogData logData) throws Exception{
+        LogEntry logEntry;
+        try {
+            logEntry = logData.getLogEntry(runtime);
+        } catch (Exception e) {
+            log.error("Cannot deserialize log entry" + logData.getGlobalAddress(), e);
+            throw e;
+        }
+
+        return logEntry;
+    }
+
+    /** Extract log entries from logData and update the Corfu Objects
+     *
+     * @param logData LogData received from Corfu server.
+     */
+    private void processLogData(ILogData logData) {
+
+        LogEntry logEntry;
+        try {
+            logEntry = deserializeLogData(logData);
+        } catch (Exception e) {
+            return;
+        }
+
+        if (logEntry.getType() == LogEntry.LogEntryType.SMR) {
+            // Just one stream, always
+            UUID streamId = logData.getStreams().iterator().next();
+            applySmrEntryToStream(streamId, (SMREntry) logEntry);
+        }
+        else if (logEntry.getType() == LogEntry.LogEntryType.MULTIOBJSMR) {
+            MultiObjectSMREntry multiObjectLogEntry = (MultiObjectSMREntry) logEntry;
+            multiObjectLogEntry.getEntryMap().forEach((stream, multiSmrEntry) -> {
+                multiSmrEntry.getSMRUpdates(stream).forEach((smrEntry) -> {
+                    applySmrEntryToStream(stream, smrEntry);
+                });
+            });
+
+        }
+    }
+
+    private void findAndSetLogHead() {
+        logHead = 0;
+    }
+
+    private void findAndSetLogTail() {
+        logTail = runtime.getSequencerView().nextToken(Collections.emptySet(), 0).getTokenValue();
+    }
+
+    /** Initialize log head and log tails
+     *
+     * If logHead and logTail has not been initialized by
+     * the user, initialize to default.
+     *
+     */
+    private void initializeHeadAndTails() {
+        if (logHead < 0) {
```

Comment:
I wasn't implying that you should rely on the sequencer. I was talking about the information flow, it seems like you can acquire both the log head and log tail from within the class, why make this configurable externally ?